internal/cli/root/mongocli/builder.go

```diff
@@ -61,11 +56,10 @@ func Builder(profile *string, argsWithoutProg []string) *cobra.Command {
 			if shouldSkipPrintNewVersion(w) {
 				return
 			}
-			opts := &BuilderOpts{
-				store: latestrelease.NewPrinter(context.Background()),
-			}
 
-			_ = opts.store.PrintNewVersionAvailable(w, version.Version, config.ToolName, config.BinName())
+			printer := latestrelease.NewPrinter(w, config.ToolName, config.BinName())
+			checker := latestrelease.NewChecker(version.Version, config.ToolName, printer)
+			_ = checker.CheckAvailable()
```

Comment:
definitely getting closer but here's my train of thought, if I see 
`printer := latestrelease.NewPrinter(w, config.ToolName, config.BinName())` that's fine, it's a printer, taking a couple of arguments
but then I see
`checker := latestrelease.NewChecker(version.Version, config.ToolName, printer)` questions come to mind, why does a "checker" need a printer? 👃 
then I see `checker.CheckAvailable()` and you lost me there what does `CheckAvailable` do, how does it use a printer, you go and see it's an interface so who knows what it does and why it needs a printer for some reason

Keep the boundaries between your functions small, `CheckAvailable` does too much, it checks the cached state, fetches from github, updates the cache state, then prints a message 

The problems of `CheckAvailable` are also how much internal knowledge it has on how to do some of those things, it directly calls a bunch of `NewSomething` inside `latestVersion` which makes the code harder to test and with a secret dependency on all those `NewSomething`, all of those dependencies should be injected, or refer to the original point, `CheckAvailable` does too much, reducing responsibilities will reduce the number of things you need to inject

Reply:
from your proposed solution I just wonder why the checker needs a hombrew loader/saver
I also wonder why checker returns a homebrew command, my high level expectations would be to 
```
// check cache -> if cache invalid -> check latest version -> if !latest version -> check hombrew -> print
```

In particular I don't think the version checker should be handling any I/O at all